### PR TITLE
feat: feature gates

### PR DESCRIPTION
**Pull request by brianp**
Description
---
This PR is for reflective purposes regarding feature gating at compile time. See https://github.com/tari-project/tari/issues/5135 for more info. This PR as it stands is to be used as an example of a network _type_ based feature gate at compile time.
It allows setting an ENVVAR for the network with conditional compilation for features. The feature sets can be imported across any crate in the project easily and then used with the `#[cfg(tari_feature_...)]` attribute macro.

I've also introduced a secondary commit for binaries to perform a quick network check against itself to ensure the intended network is supported by the binary. This check is made non-invasive (not conditionally compiling possible network) as that method of check ends up being much more tedious.

Motivation and Context
---
We want some features on some networks, but maybe not all.

How Has This Been Tested?
---
Manually

What process can a PR reviewer use to test or verify this change?
---
Compile a binary with a set network
`TARI_NETWORK=nextnet cargo build --release --bin tari_base_node`

then run the binary with a test network as a parameter:

`./tari_base_node --network esme` 

and receive an error:

```
The application exited because of an internal network error. The network esmeralda is invalid for this binary built for NextNet
```

*Please note*

That running the binary will cause the build to default to the test network _always_. Meaning this will not work:

```
$ TARI_NETWORK=nextnet cargo build --release --bin tari_base_node
$ cargo run --release --bin tari_base_node --network nextnet
```

The second command `cargo run` will re-build the binary, and without the `TARI_NETWORK` env set the binary will default to a test binary.

In development if you want to call run and connect to a non test network you need to also pass the TARI_NETWORK envvar instead of the `--network` flag. 
`TARI_NETWORK=nextnet cargo run --bin tari_base_node --release -- --network nextnet`

Breaking Changes
---

- [x] None

### CODE REVIEW

File: applications/tari_app_utilities/src/network_check.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        },
+        Target::NextNet => match network {
+            Network::NextNet => Ok(()),
+            _ => Err(NetworkCheckError::NextNetBinary(network)),
+        },
+        Target::TestNet => match network {
+            Network::LocalNet | Network::Igor | Network::Esmeralda => Ok(()),
+            _ => Err(NetworkCheckError::TestNetBinary(network)),
+        },
+    }
```
Comment (CjS77) [inline applications/tari_app_utilities/src/network_check.rs]:
```suggestion
    match (TARGET_NETWORK, network) {
        (Target::MainNet, Network::MainNet | Network::StageNet) => Ok(()),
        (Target::MainNet,   _) => Err(NetworkCheckError::MainNetBinary(network)),
        (Target::NextNet,  Network::NextNet) => Ok(()),
        (Target::NextNet, _) => Err(NetworkCheckError::NextNetBinary(network)),
        (Target::TestNet,   Network::LocalNet | Network::Igor | Network::Esmeralda) => Ok(()),
        (Target::TestNet, _) => Err(NetworkCheckError::TestNetBinary(network)),
        },
    }
```

A lot more readable I think you'll find

File: common/tari_features/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+mod feature;
+pub mod resolver;
+mod status;
+
+pub use feature::Feature;
+pub use status::Status;
+
```
Comment (CjS77) [inline common/tari_features/src/lib.rs:29]: Some docs here on when / how / why to maintain this list would be highly useful.

**Review by CjS77 — COMMENTED** [review]
Looks good. Some minor readability and maintenance suggestions

File: common/tari_features/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "tari_features"
+description = "Compilable features for Tari applications"
+authors = ["The Tari Development Community"]
+repository = "https://github.com/tari-project/tari"
+homepage = "https://tari.com"
+readme = "README.md"
+license = "BSD-3-Clause"
+version = "0.0.1"
```
Comment (SWvheerden) [inline common/tari_features/Cargo.toml]: This is probably up for discussion, but I think we should version match here with the rest of the Tari code base. 
at least for now.

File: applications/tari_app_utilities/Cargo.toml [hunk 4]
```diff
@@ -22,3 +23,4 @@ thiserror = "^1.0.26"
 
 [build-dependencies]
 tari_common = { path = "../../common", features = ["build", "static-application-info"] }
+tari_features = { version = "0.0.1", path = "../../common/tari_features"}
```
Comment (SWvheerden) [inline applications/tari_app_utilities/Cargo.toml]: is there a reason for including it twice with the same feature set?

File: applications/tari_app_utilities/src/network_check.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    TestNetBinary(Network),
+}
+
+impl From<NetworkCheckError> for ExitError {
+    fn from(err: NetworkCheckError) -> Self {
+        Self::new(ExitCode::NetworkError, err)
+    }
+}
+
+#[cfg(tari_network_mainnet)]
```
Comment (SWvheerden) [inline applications/tari_app_utilities/src/network_check.rs:46]: Can we add a default to this?
So if the user does not specify a network, it defaults to mainnet

File: common/tari_features/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+pub const FEATURE_LIST: [Feature; 0] = [
+    // Feature::new(
+    //     "example_feature_name",
+    //     "example description",
+    //     None,
+    //     Status::New,
+    // ),
```
Comment (SWvheerden) [inline common/tari_features/src/lib.rs]: remove?

**Review by SWvheerden — COMMENTED** [review]
This looks good
The only thing I would add is a default network, so if the user does not select a network, it will default to mainnet, or perhaps development as building from dev that's the most likely? But this is open to discussion

**Review by SWvheerden — COMMENTED** [review]
We need to update the readme with all this as well.

File: common/tari_features/Cargo.toml (hunk 3 above)
Comment (brianp) [reply common/tari_features/Cargo.toml]: That makes sense to me.

File: applications/tari_app_utilities/Cargo.toml (hunk 4 above)
Comment (brianp) [reply applications/tari_app_utilities/Cargo.toml]: I was thinking it wasn't available in the build deps unless specifically defined, although I'll double check it's needed.

File: applications/tari_app_utilities/src/network_check.rs (hunk 5 above)
Comment (brianp) [reply applications/tari_app_utilities/src/network_check.rs:46]: There is a default defined as the Target::TestNet. Anytime we do compiled compilation we have to ensure something will be present. For all the feature gating TestNet is the default compilation path.

`#[cfg(all(not(tari_network_mainnet), not(tari_network_nextnet)))]`

File: applications/tari_app_utilities/Cargo.toml (hunk 4 above)
Comment (brianp) [reply applications/tari_app_utilities/Cargo.toml]: Validated. For use in build.rs it _is_ required to be included in both categories.

**Review by SWvheerden — APPROVED** [review]

**Comment by SWvheerden** [comment]
Ack

**Pull request merged**